cache/demux.go

```diff
@@ -166,6 +166,31 @@ func (d *demux) Purge() {
 	d.laggingWatchers = make(map[*watcher]int64)
 }
 
+// Compact is called when etcd reports a compaction at compactRev to rebase history;
+// it keeps provably-too-old watchers for later cancellation, stops others, and clients should resubscribe.
+func (d *demux) Compact(compactRev int64) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.history.RebaseHistory()
+
+	for w, next := range d.activeWatchers {
+		if next != 0 && next <= compactRev {
+			delete(d.activeWatchers, w)
+			d.laggingWatchers[w] = next
+		} else {
+			w.Stop()
+			delete(d.activeWatchers, w)
+		}
+	}
+	for w, next := range d.laggingWatchers {
+		if next != 0 && next > compactRev {
+			continue
+		}
+		w.Stop()
+		delete(d.laggingWatchers, w)
```

Comment:
Lagging watchers should also be informed about being compacted. Why do we just drop them?